fpmsyncd/routesync.cpp

```diff
@@ -1222,3 +1269,183 @@ string RouteSync::getNextHopWt(struct rtnl_route *route_obj)
+
+    return sendOffloadReply(nlmsg_hdr(nlMsg.get()));
+}
+
+void RouteSync::setSuppressionEnabled(bool enabled)
+{
+    SWSS_LOG_ENTER();
+
+    m_isSuppressionEnabled = enabled;
+
+    SWSS_LOG_NOTICE("Pending routes suppression is %s", (m_isSuppressionEnabled ? "enabled": "disabled"));
+}
+
+void RouteSync::onRouteResponse(const std::string& key, const std::vector<FieldValueTuple>& fieldValues)
+{
+    IpPrefix prefix;
+    std::string vrfName;
+    std::string protocol;
+
+    bool isSetOperation{false};
+    bool isSuccessReply{false};
+
+    if (!isSuppressionEnabled())
+    {
+        return;
+    }
+
+    auto colon = key.find(':');
+    if (colon != std::string::npos && key.substr(0, colon).find(VRF_PREFIX) != std::string::npos)
+    {
+        vrfName = key.substr(0, colon);
+        prefix = IpPrefix{key.substr(colon + 1)};
+    }
+    else
+    {
+        prefix = IpPrefix{key};
+    }
+
+    for (const auto& fieldValue: fieldValues)
+    {
+        std::string field = fvField(fieldValue);
+        std::string value = fvValue(fieldValue);
+
+        if (field == "err_str")
+        {
+            isSuccessReply = (value == "SWSS_RC_SUCCESS");
+        }
+        else if (field == "protocol")
+        {
+            // If field "protocol" is present in the field values then
+            // it is a SET operation. This field is absent only if we are
+            // processing DEL operation.
+            isSetOperation = true;
+            protocol = value;
+        }
+    }
+
+    if (!isSetOperation)
+    {
+        SWSS_LOG_DEBUG("Received response for prefix %s(%s) deletion, ignoring ",
+            prefix.to_string().c_str(), vrfName.c_str());
+        return;
+    }
+
+    auto routeObject = makeUniqueWithDestructor(rtnl_route_alloc(), rtnl_route_put);
+    auto dstAddr = makeNlAddr(prefix);
+
+    rtnl_route_set_dst(routeObject.get(), dstAddr.get());
+
+    auto proto = rtnl_route_str2proto(protocol.c_str());
+    if (proto < 0)
+    {
+        proto = swss::to_uint<uint8_t>(protocol);
+    }
+
+    rtnl_route_set_protocol(routeObject.get(), static_cast<uint8_t>(proto));
+    rtnl_route_set_family(routeObject.get(), prefix.isV4() ? AF_INET : AF_INET6);
+
+    unsigned int vrfIfIndex = 0;
+    if (!vrfName.empty())
+    {
+        auto* link = m_linkCache.getLinkByName(vrfName.c_str());
+        if (!link)
+        {
+            SWSS_LOG_DEBUG("Failed to find VRF when constructing response message for prefix %s(%s). "
+                "This message is probably outdated", prefix.to_string().c_str(),
+                vrfName.c_str());
+            return;
+        }
+        vrfIfIndex = rtnl_link_get_ifindex(link);
+    }
+
+    rtnl_route_set_table(routeObject.get(), vrfIfIndex);
+
+    unsigned int flags = 0;
+
+    if (isSuccessReply)
+    {
+        flags |= RTM_F_OFFLOAD;
+    }
+
+    // Mark route as OFFLOAD
+    rtnl_route_set_flags(routeObject.get(), flags);
+
+    if (!sendOffloadReply(routeObject.get()))
```

Comment:
When there are multiple route updates for the given route, how do we make sure that the response to sent for the right route-update?

Reply:
The constructed RTM_NEWROUTE message contains prefix + vrf + protocol of the route